File: src/main/java/ru/melonhell/bulletphysics/HitboxRender.java
package ru.melonhell.bulletphysics;

import com.jme3.bullet.objects.PhysicsRigidBody;
import com.jme3.math.Quaternion;
import com.jme3.math.Vector3f;
import lombok.experimental.UtilityClass;
import org.bukkit.*;
import ru.melonhell.bulletphysics.bullet.collision.body.element.PhysicsElement;
import ru.melonhell.bulletphysics.bullet.collision.body.shape.Triangle;
import ru.melonhell.bulletphysics.init.BulletPhysicsPlugin;
import ru.melonhell.bulletphysics.utils.math.MathShit;

import java.util.Arrays;
import java.util.List;

@UtilityClass
public class HitboxRender {

    public void render(PhysicsElement physicsElement, World world) {
        Bukkit.getScheduler().runTaskAsynchronously(BulletPhysicsPlugin.getPlugin(BulletPhysicsPlugin.class), () -> renderSync(physicsElement, world));
    }
    public void renderSync(PhysicsElement physicsElement, World world) {
        PhysicsRigidBody rigidBody = physicsElement.getRigidBody();
        Vector3f position = rigidBody.getMotionState().getLocation(new Vector3f());
        Quaternion orientation = rigidBody.getMotionState().getOrientation(new Quaternion());
        List<Triangle> triangles = physicsElement.getMinecraftShape().getTriangles(Quaternion.IDENTITY);
        for (Triangle triangle : triangles) {
            triangle(triangle, world, orientation, position, false);
        }
    }

    public void triangle(Triangle triangle, World world, Quaternion orientation, Vector3f position, boolean lines) {
        List<Vector3f> vertices = Arrays.stream(triangle.transform(orientation).getVertices()).map(vector3f -> vector3f.add(position)).toList();
        for (int i = 0; i < 3; i++) {
            Vector3f v1 = vertices.get(i);
            Vector3f v2 = vertices.get((i + 1) % 3);
            if (lines) line(v1, v2, 3, world);
            particle(new Location(world, v1.x, v1.y, v1.z), Color.RED);
        }
    }

    public void line(Vector3f v1, Vector3f v2, int count, World world) {
        for (int i = 1; i < count; i++) {
            Vector3f lerp = MathShit.lerp(v1, v2, (float) i / count);
            particle(new Location(world, lerp.x, lerp.y, lerp.z), Color.YELLOW);
        }
    }

    public void particle(Location location, Color color) {
        location.getWorld().spawnParticle(Particle.REDSTONE, location, 1, 0, 0, 0, new Particle.DustOptions(color, 1));
    }
}

File: src/main/java/ru/melonhell/bulletphysics/HitboxRender.kt
package ru.melonhell.bulletphysics

import com.jme3.math.Quaternion
import com.jme3.math.Vector3f
import org.bukkit.*
import org.bukkit.Particle.DustOptions
import org.bukkit.plugin.java.JavaPlugin
import ru.melonhell.bulletphysics.bullet.collision.body.element.PhysicsElement
import ru.melonhell.bulletphysics.bullet.collision.body.shape.Triangle
import ru.melonhell.bulletphysics.init.BulletPhysicsPlugin
import ru.melonhell.bulletphysics.utils.math.MathShit
import java.util.*

object HitboxRender {
    @JvmStatic
    fun render(physicsElement: PhysicsElement, world: World) {
        Bukkit.getScheduler().runTaskAsynchronously(JavaPlugin.getPlugin(
            BulletPhysicsPlugin::class.java
        ), Runnable { renderSync(physicsElement, world) })
    }

    @JvmStatic
    fun renderSync(physicsElement: PhysicsElement, world: World) {
        val rigidBody = physicsElement.rigidBody
        val position = rigidBody.motionState.getLocation(Vector3f())
        val orientation = rigidBody.motionState.getOrientation(Quaternion())
        val triangles = physicsElement.minecraftShape.getTriangles(Quaternion.IDENTITY)
        for (triangle in triangles) {
            triangle(triangle, world, orientation, position, false)
        }
    }

    @JvmStatic
    fun triangle(triangle: Triangle, world: World, orientation: Quaternion, position: Vector3f, lines: Boolean) {
        val vertices = Arrays.stream(triangle.transform(orientation).vertices)
            .map { vector3f: Vector3f -> vector3f.add(position) }
            .toList()
        for (i in 0..2) {
            val v1 = vertices[i]
            val v2 = vertices[(i + 1) % 3]
            if (lines) line(v1, v2, 3, world)
            particle(Location(world, v1.x.toDouble(), v1.y.toDouble(), v1.z.toDouble()), Color.RED)
        }
    }

    @JvmStatic
    fun line(v1: Vector3f, v2: Vector3f, count: Int, world: World) {
        for (i in 1 until count) {
            val lerp = MathShit.lerp(v1, v2, i.toFloat() / count)
            particle(Location(world, lerp.x.toDouble(), lerp.y.toDouble(), lerp.z.toDouble()), Color.YELLOW)
        }
    }

    @JvmStatic
    fun particle(location: Location, color: Color?) {
        location.world.spawnParticle(
            Particle.REDSTONE, location, 1, 0.0, 0.0, 0.0, DustOptions(
                color!!, 1f
            )
        )
    }
}